File: app/src/main/java/fr/project/mlignereux/cv/view/fragment/ContactFragment.java
package fr.project.mlignereux.cv.view.fragment;

import android.Manifest;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import fr.project.mlignereux.cv.model.Contact;

public class ContactFragment extends Fragment implements ActivityCompat.OnRequestPermissionsResultCallback {

    private final static String TAG = "ContactFragment";

    private Boolean isFabOpen = false;
    private FloatingActionButton mFab, mFab1, mFab2;
    private Animation fab_open,fab_close,rotate_forward,rotate_backward;
    private TextView mEmailTv, mCallTv;

    private DatabaseReference mDatabase;

    private Context mContext;

    private static final int CALL_PERMISSION = 1;
    private String callNumber;

    public ContactFragment() {
        // Required empty public constructor
    }

    public static ContactFragment newInstance(String reference, String title, String subtitle, String description) {

        ContactFragment fragment = new ContactFragment();
        Bundle args = new Bundle();
        args.putString("REFERENCE", reference);
        args.putString("TITLE", title);
        args.putString("SUBTITLE", subtitle);
        args.putString("DESCRIPTION", description);
        fragment.setArguments(args);

        return fragment;
    }


    public String getReference(){
        return getArguments().getString("REFERENCE");
    }

    public String getTitle(){
        return getArguments().getString("TITLE");
    }

    public String getSubtitle() { return getArguments().getString("SUBTITLE");}

    public String getDescription() { return getArguments().getString("DESCRIPTION"); }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mContext= getContext().getApplicationContext();
        mDatabase = FirebaseDatabase.getInstance().getReference();

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        final View v =  inflater.inflate(fr.project.mlignereux.cv.R.layout.fragment_contact, container, false);

        mFab = (FloatingActionButton) v.findViewById(fr.project.mlignereux.cv.R.id.fab);
        mFab1 = (FloatingActionButton) v.findViewById(fr.project.mlignereux.cv.R.id.fab1);
        mFab2 = (FloatingActionButton) v.findViewById(fr.project.mlignereux.cv.R.id.fab2);
        mCallTv = (TextView) v.findViewById(fr.project.mlignereux.cv.R.id.tv_call);
        mEmailTv = (TextView) v.findViewById(fr.project.mlignereux.cv.R.id.tv_email);

        fab_open = AnimationUtils.loadAnimation(mContext, fr.project.mlignereux.cv.R.anim.fab_open);
        fab_close = AnimationUtils.loadAnimation(mContext, fr.project.mlignereux.cv.R.anim.fab_close);
        rotate_forward = AnimationUtils.loadAnimation(mContext, fr.project.mlignereux.cv.R.anim.rotate_forward);
        rotate_backward = AnimationUtils.loadAnimation(mContext, fr.project.mlignereux.cv.R.anim.rotate_backward);

        ((AppCompatActivity)getActivity()).getSupportActionBar().setTitle(getTitle());

        mFab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                animateFAB();
            }
        });

        final TextView stateTextView = (TextView)v.findViewById(fr.project.mlignereux.cv.R.id.tv_contact_state);
        final TextView bornDateTextView = (TextView)v.findViewById(fr.project.mlignereux.cv.R.id.tv_contact_born_date);
        final TextView mailTextView = (TextView)v.findViewById(fr.project.mlignereux.cv.R.id.tv_contact_mail);
        final TextView phoneTextView = (TextView)v.findViewById(fr.project.mlignereux.cv.R.id.tv_contact_phone);
        final TextView addressTextView = (TextView)v.findViewById(fr.project.mlignereux.cv.R.id.tv_contact_address);
        final TextView driversLicenseTextView = (TextView)v.findViewById(fr.project.mlignereux.cv.R.id.tv_contact_drivers_license);
        final TextView linkedinTextView = (TextView)v.findViewById(fr.project.mlignereux.cv.R.id.tv_contact_linkedin);
        final TextView viadeoTextView = (TextView)v.findViewById(fr.project.mlignereux.cv.R.id.tv_contact_viadeo);
        final ImageView viadeoImageView = (ImageView)v.findViewById(fr.project.mlignereux.cv.R.id.iv_contact_viadeo);
        final ImageView linkedinImageView = (ImageView)v.findViewById(fr.project.mlignereux.cv.R.id.iv_contact_linkedin);


        mDatabase.child(getReference()).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {

                final Contact contact = dataSnapshot.getValue(Contact.class);

                stateTextView.setText(contact.getState());
                bornDateTextView.setText(contact.getBornDate());
                mailTextView.setText(contact.getEmail());
                phoneTextView.setText(contact.getPhone());
                addressTextView.setText(contact.getAddress());
                driversLicenseTextView.setText(contact.getDriversLicense());
                linkedinTextView.setText(contact.getLinkedin());

                viadeoTextView.setText(contact.getViadeo());


                Glide.with(mContext).load(contact.getLinkedinIcon()).skipMemoryCache( true ).into(linkedinImageView);
                Glide.with(mContext).load(contact.getViadeoIcon()).skipMemoryCache( true ).into(viadeoImageView);

                callNumber = contact.getPhone();

                call(callNumber);

                sendMail(contact.getEmail());



            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });



        return v;
    }

    private void call(final String number){
        mFab1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                if(ContextCompat.checkSelfPermission(mContext, Manifest.permission.CALL_PHONE) != PackageManager.PERMISSION_GRANTED){
                    Log.d(TAG, "checkpermission");

                    if(ActivityCompat.shouldShowRequestPermissionRationale(getActivity(), Manifest.permission.CALL_PHONE)){
                        Log.d(TAG, "shouldRequest");

                        // Show an explanation to the user *asynchronously* -- don't block
                        // this thread waiting for the user's response! After the user
                        // sees the explanation, try again to request the permission.

                        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

                        builder.setMessage("Vous devez autoriser l'application à utiliser votre appareil pour téléphoner");

                        builder.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                requestPermissions(new String[]{Manifest.permission.CALL_PHONE}, CALL_PERMISSION);
                            }
                        });
                        builder.setNegativeButton("Annuler", new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                dialog.dismiss();
                            }
                        });

                        AlertDialog dialog = builder.create();

                        dialog.show();

                    }else{

                        // No explanation needed, we can request the permission.
                        requestPermissions(new String[]{Manifest.permission.CALL_PHONE}, CALL_PERMISSION);
                    }
                }else{
                    Intent callIntent = new Intent(Intent.ACTION_CALL);
                    callIntent.setData(Uri.parse("tel:" + number));
                    startActivity(callIntent);
                }


            }
        });
    }

    private void sendMail(final String email){

        mFab2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent i = new Intent(Intent.ACTION_SEND);
                i.setType("message/rfc822");
                i.putExtra(Intent.EXTRA_EMAIL  , new String[]{email});

                try {
                    startActivity(Intent.createChooser(i, "Send mail..."));
                } catch (android.content.ActivityNotFoundException ex) {
                    Toast.makeText(getActivity(), "There are no email clients installed.", Toast.LENGTH_SHORT).show();
                }
            }
        });

    }

    private void animateFAB(){

        if(isFabOpen){

            mFab.startAnimation(rotate_backward);
            mFab1.startAnimation(fab_close);
            mFab2.startAnimation(fab_close);
            mCallTv.startAnimation(fab_close);
            mEmailTv.startAnimation(fab_close);
            mFab1.setClickable(false);
            mFab2.setClickable(false);
            isFabOpen = false;

        } else {

            mFab.startAnimation(rotate_forward);
            mFab1.startAnimation(fab_open);
            mFab2.startAnimation(fab_open);
            mCallTv.startAnimation(fab_open);
            mEmailTv.startAnimation(fab_open);
            mFab1.setClickable(true);
            mFab2.setClickable(true);
            isFabOpen = true;

        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull  String permissions[], @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        switch (requestCode) {

            case CALL_PERMISSION: {

                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {

                    Intent callIntent = new Intent(Intent.ACTION_CALL);
                    callIntent.setData(Uri.parse("tel:" + callNumber));
                    startActivity(callIntent);

                }
            }


        }
    }


}

File: app/src/main/kotlin/fr/project/mlignereux/cv/view/fragment/ContactFragment.kt
package fr.project.mlignereux.cv.view.fragment

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.firebase.database.*
import dagger.android.support.DaggerFragment
import fr.project.mlignereux.R
import fr.project.mlignereux.cv.model.Contact
import fr.project.mlignereux.cv.view.ImageLoader
import javax.inject.Inject

class ContactFragment : DaggerFragment(), ActivityCompat.OnRequestPermissionsResultCallback {

    @Inject
    lateinit var imageLoader: ImageLoader
    @Inject
    lateinit var database: DatabaseReference

    private var isFabOpen: Boolean = false
    private lateinit var fab: FloatingActionButton
    private lateinit var fab1: FloatingActionButton
    private lateinit var fab2: FloatingActionButton
    private lateinit var fabOpen: Animation
    private lateinit var fabClose: Animation
    private lateinit var rotateForward: Animation
    private lateinit var rotateBackward: Animation
    private lateinit var emailTv: TextView
    private lateinit var callTv: TextView

    private lateinit var callNumber: String

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val view = inflater.inflate(R.layout.fragment_contact, container, false)

        fab = view.findViewById(R.id.fab) as FloatingActionButton
        fab1 = view.findViewById(R.id.fab1) as FloatingActionButton
        fab2 = view.findViewById(R.id.fab2) as FloatingActionButton
        callTv = view.findViewById(R.id.tv_call) as TextView
        emailTv = view.findViewById(R.id.tv_email) as TextView

        fabOpen = AnimationUtils.loadAnimation(requireContext(), R.anim.fab_open)
        fabClose = AnimationUtils.loadAnimation(requireContext(), R.anim.fab_close)
        rotateForward = AnimationUtils.loadAnimation(requireContext(), R.anim.rotate_forward)
        rotateBackward = AnimationUtils.loadAnimation(requireContext(), R.anim.rotate_backward)

        (activity as AppCompatActivity).supportActionBar?.title = getString(R.string.contact_title)

        fab.setOnClickListener { animateFAB() }

        val stateTextView = view.findViewById(R.id.tv_contact_state) as TextView
        val bornDateTextView = view.findViewById(R.id.tv_contact_born_date) as TextView
        val mailTextView = view.findViewById(R.id.tv_contact_mail) as TextView
        val phoneTextView = view.findViewById(R.id.tv_contact_phone) as TextView
        val addressTextView = view.findViewById(R.id.tv_contact_address) as TextView
        val driversLicenseTextView = view.findViewById(R.id.tv_contact_drivers_license) as TextView
        val linkedinTextView = view.findViewById(R.id.tv_contact_linkedin) as TextView
        val viadeoTextView = view.findViewById(R.id.tv_contact_viadeo) as TextView
        val viadeoImageView = view.findViewById(R.id.iv_contact_viadeo) as ImageView
        val linkedinImageView = view.findViewById(R.id.iv_contact_linkedin) as ImageView

        database.child(getString(R.string.contact_bundle)).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {

                val contact = dataSnapshot.getValue(Contact::class.java) ?: Contact()

                stateTextView.text = contact.state
                bornDateTextView.text = contact.bornDate
                mailTextView.text = contact.email
                phoneTextView.text = contact.phone
                addressTextView.text = contact.address
                driversLicenseTextView.text = contact.driversLicense
                linkedinTextView.text = contact.linkedin
                viadeoTextView.text = contact.viadeo

                imageLoader.loadImage(contact.linkedinIcon, linkedinImageView)
                imageLoader.loadImage(contact.viadeoIcon, viadeoImageView)

                callNumber = contact.phone ?: ""
                call(callNumber)

                sendMail(contact.email)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
        return view
    }

    private fun call(number: String?) {
        fab1.setOnClickListener {
            if (ContextCompat.checkSelfPermission(requireContext(), Manifest.permission.CALL_PHONE) != PackageManager.PERMISSION_GRANTED) {

                if (ActivityCompat.shouldShowRequestPermissionRationale(requireActivity(), Manifest.permission.CALL_PHONE)) {

                    val builder = AlertDialog.Builder(requireContext())
                    builder.setMessage("Vous devez autoriser l'application à utiliser votre appareil pour téléphoner")
                    builder.setPositiveButton("Ok") { dialog, id ->
                        requestPermissions(arrayOf(Manifest.permission.CALL_PHONE), CALL_PERMISSION)
                    }
                    builder.setNegativeButton("Annuler") { dialog, id -> dialog.dismiss() }

                    val dialog = builder.create()
                    dialog.show()

                } else {
                    requestPermissions(arrayOf(Manifest.permission.CALL_PHONE), CALL_PERMISSION)
                }
            } else {
                val callIntent = Intent(Intent.ACTION_CALL)
                callIntent.data = Uri.parse("tel:$number")
                startActivity(callIntent)
            }
        }
    }

    private fun sendMail(email: String?) {

        fab2.setOnClickListener {
            val i = Intent(Intent.ACTION_SEND)
            i.type = "message/rfc822"
            i.putExtra(Intent.EXTRA_EMAIL, arrayOf(email.orEmpty()))
            try {
                startActivity(Intent.createChooser(i, "Send mail..."))
            } catch (ex: android.content.ActivityNotFoundException) {
                Toast.makeText(activity, "There are no email clients installed.", Toast.LENGTH_SHORT).show()
            }
        }

    }

    private fun animateFAB() {
        if (isFabOpen) {
            fab.startAnimation(rotateBackward)
            fab1.startAnimation(fabClose)
            fab2.startAnimation(fabClose)
            callTv.startAnimation(fabClose)
            emailTv.startAnimation(fabClose)
            fab1.isClickable = false
            fab2.isClickable = false
            isFabOpen = false
        } else {
            fab.startAnimation(rotateForward)
            fab1.startAnimation(fabOpen)
            fab2.startAnimation(fabOpen)
            callTv.startAnimation(fabOpen)
            emailTv.startAnimation(fabOpen)
            fab1.isClickable = true
            fab2.isClickable = true
            isFabOpen = true
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        when (requestCode) {
            CALL_PERMISSION -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    val callIntent = Intent(Intent.ACTION_CALL)
                    callIntent.data = Uri.parse("tel:$callNumber")
                    startActivity(callIntent)
                }
            }
        }
    }

    companion object {
        private const val CALL_PERMISSION = 1
    }

}
